### DISCUSSION TITLE
Security for platform services

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [fastapi/fastapi#9375] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Example code not available
```

### Description

Assume a dockerized FastApi service (call it Zippy) which is available from a cloud platform.

Organizations register for Zippy with login credentials.

An organization creates one or more Zippy-based services. 

Users access a Zippy service anonymously or login.

Need FastApi to handle the security access for an organization and the individual services ie.
```
Cloud
    Zippy - a docker'ized FastApi-based service 
        Org-1
            zippy-service-11
            zippy-service-12
            ...
            zippy-service-1n
        Org-2
            zippy-service-21
            zippy-service-22
            ...
            zippy-service-2p
        ...
        Org-m
            zippy-service-m1
            zippy-service-m2
            ...
            zippy-service-mr
```
I've read and understood the FastApi security documentation. 

What is the guidance to achieve the above with FastApi security? Thanks

### Operating System

Windows

### FastAPI Version

0.95

### Python Version

3.10.8

**Comment by followben** [comment]
If I understand you correctly, you're asking how to do common authentication across multiple fastapi-based services?

For a prototype, the "simplest thing that works" with fastapi alone might be to [follow the simple auth example](https://fastapi.tiangolo.com/tutorial/security/simple-oauth2/), but use a read-only shared datastore for each service's dependency like `get_current_user` in the example. If you want to keep your services data layers isolated from each other, they could in turn make a HTTP call to a common fastapi that does the authentication.

A more sophisticated and real-world implementation would be to put your services behind a reverse proxy or API gateway which in turn does the authentication (so the services never see the request unless it's already authenticated). For example, if you're on AWS and using their APIGateway product, you could deploy a separate fastAPI "authentication" service as a [lambda authorizer](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-lambda-authorizer.html).

Another common pattern (that can also be combined with either approach above) is to have a service that vends bearer tokens that contain and validate auth state, such as a JWT or Jose. You can build that yourself or use an SaaS like Auth0 or Cognito, or run an open source solution like Keycloak. But for most use cases this is likely overkill and you'd be [better off managing your own session tokens](https://redis.com/blog/json-web-tokens-jwt-are-dangerous-for-user-sessions/).

**Reply by dineshbvadhia** [reply]
Hope this clarifies the need: Organizations install Zippy (a fastapi-based service) to run on a server from a docker container residing on a cloud platform. The fastapi-based service uses jwt authentication to authenticate organizations and users. The question becomes, is one secret stored in say, environment variables in one docker image, sufficient for each Org(k)-Service(m)-Users(n) server instance?

**Reply by followben** [reply, marked as answer]
Well I wouldn't store secrets in docker env - better pass via a mount or retrieve them dynamically from a trusted provider. [Lots](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/secrets-app.html) [of](https://kubernetes.io/docs/concepts/configuration/secret/) [solutions](https://snyk.io/blog/keeping-docker-secrets-secure/) [for this problem](https://www.hashicorp.com/resources/securing-container-secrets-vault) depending on how you run your containers. If you're asking because you want to store the key for decoding your JWTs, then that's the way to go.
